### DISCUSSION TITLE
Misorientation-dependent gb energy and gb mobility anisotropy

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#20570] [Q&A Modules: Phase field] [answered]
Dear MOOSE Experts,

Here, I created a [pull requests](https://github.com/idaholab/moose/pull/20439) expecting to introduce grain boundary energy and grain boundary mobility anisotropy due to grain boundary misorientation in moose's grain growth simulation. But here are some questions to ask you.

When I code the `ComputePolycrystalGBAnisotropy` material object to get the grain misorientation `_delta_theta` at each quadrature point, based on the Euler angles from file. 

Then, [GBAnisotropyMisorientation](https://github.com/PengWei97/moose/blob/numAdjGrain_gg/modules/phase_field/src/materials/GBAnisotropyMisorientation.C) was created to compute the model parameters $\kappa(\theta), L(\theta), \gamma(\theta)$ and m in the multi-order phase field model, which refer to `GBAnisotropy` and [B. Misorientation dependence in Nele Moelans' PRB](https://link.aps.org/doi/10.1103/PhysRevB.78.024113). 

The only difference between `GBAnisotropyMisorientation` and `GBAnisotropy` is that the gb energy $\sigma_{gb}$ and gb mobility $\mu_{gb}$ output by `GBAnisotropyMisorientation` are not based on the array[num_op, num_op] given in the file, but are calculated based on the misorientation $\theta$ given by this quadrature point.

Next, when I create an example, [anisotropic_GB_100_N0load.i](https://github.com/PengWei97/luwu/blob/main/examples/phasefield/GBAnisotropy/anisotropic_GB_100_N0load.i), based on `grain_growth_2D_graintracker.i` and add the following code in the material module,

```bash
[Materials]
  [./CuGrGr]
    # Material properties
    type = GBAnisotropyMisorientation # GBAnisotropyEvolutionBase GBEvolution GBAnisotropyMisorientation
    T = 450 # Constant temperature of the simulation (for mobility calculation)
    wGB = 14 # Width of the diffuse GB
    GBmob0 = 2.5e-6 #m^4(Js) for copper from Schoenfelder1997
    Q = 0.23 #eV for copper from Schoenfelder1997
    GBenergy = 0.708 #J/m^2 from Schoenfelder1997
    outputs = my_exodus
    # output_properties = 'M_GB GBenergy' 
  [../]
  [./GBMisorientation]
    type = ComputePolycrystalGBAnisotropy
    grain_tracker = grain_tracker
    euler_angle_provider = euler_angle_file
    outputs = my_exodus
    output_properties = delta_theta
  [../]
[]
```
But here I have two problems,

First, when I set $\sigma_{gb}(\theta)$ and $\mu_{gb}(\theta)$ are based on the expressions in the [jianbao gao's article](https://www.sciencedirect.com/science/article/pii/S1359645421008697) as follows,
$$
\sigma(\theta)=\sigma_{H A B} \frac{\theta}{\theta_{H A B}}\left[1-\ln \left(\frac{\theta}{\theta_{H A B}}\right)\right]
$$

$$
\mu(\theta)=\mu_{\mathrm{HAB}}\left(1-\exp \left[-B\left(\frac{\theta}{\hat{\theta}_{\mathrm{HAB}}}\right)^{n}\right]\right)
$$

The calculation does not converge directly, as shown below

```log
Solve failed, cutting timestep.

Time Step 1, time = 1e-13, dt = 1e-13
Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0
[31m Solve Did NOT Converge![39m
Aborting as solve did not converge
```

Second, when I refer to the setting method in the [Ying Liu's article](https://www.mdpi.com/1996-1944/12/24/4048), add two constants to the above two formulas, and put $\sigma_{gb,max}(\delta \theta) / \sigma_{gb,min}(\delta \theta)$ and $\mu_{gb.max}(\delta \theta) / \mu_{gb, min}(\delta \theta)$ set to 5 can be calculated.

$$
\sigma(\theta)=4\sigma_{H A B} \frac{\theta}{\theta_{H A B}}\left[1-\ln \left(\frac{\theta}{\theta_{H A B}}\right)\right]+\sigma_{H A B}
$$

$$
\mu(\theta)=4\mu_{\mathrm{HAB}}\left(1-\exp \left[-B\left(\frac{\theta}{\hat{\theta}_{\mathrm{HAB}}}\right)^{n}\right]\right)+\mu_{\mathrm{HAB}}
$$

The figure below is the output material parameters,
![image](https://user-images.githubusercontent.com/56183775/158504227-3f6c5d8e-24ce-4d48-ba36-c63a7b006f9f.png)

By the way, in the phase field model adopted in jianbao gao's article and Ying Liu's article, grain boundary energy and grain boundary mobility are directly related to the phase field free energy functional. In Nele Moelans' PRB, the phase field model we adopt is indirectly related to these two parameters.

However, the migration of gb during the simulation process is very strange. Specifically, there are three points: a) high-angle grain boundaries do not have smooth grain boundaries, and b) inappropriately, the movement of grain boundaries is somewhat similar to the reciprocation of water waves movement, c) The width of the grain boundary does not maintain a constant size, please watch the gif animation I uploaded for details,
![GBAnotropic](https://user-images.githubusercontent.com/56183775/158504386-39b8a199-c5ee-42f2-8596-32a8d43f4a7e.gif)

So what I want to ask is if my modeling way of setting high grain boundary energy and grain boundary mobility at the grain boundary is itself a mistake. In other words, how should the grain boundary energy and grain boundary mobility anisotropy due to grain boundary misorientation be added to the multi-order parameter phase field model developed by Nele Moelans properly constructed, and expected to reproduce in Ying Liu's paper phenomenon of abnormal grain growth.

Any suggestions or recommendations to fix these problems would be greatly appreciated.
Thank you
Wei Peng

**Comment by GiudGiud** [comment, marked as answer]
@SudiptaBiswas @laagesen

**Reply by laagesen** [reply]
It is definitely expected that when you chance GB energy the GB width will also change in the Moelans 2008 grain growth model. The lack of smoothness in the grain boundaries and the jerky motion may be caused by the fact that you need to add more elements through the interfaces to make sure that you still are adequately resolving the diffuse interfaces. Make sure there are still about 3 elements through the interface.

**Reply by PengWei97** [reply]
Hi @laagesen, thank you very much for your reply.
The problem described above was solved by replacing the formula about the misorientation with the following form,
![image](https://user-images.githubusercontent.com/56183775/162615039-d3415a37-09c1-4879-839d-ea9870df10e9.png)

This may be a numerical problem, and the input grain boundary energy and grain boundary mobility cannot be set to 0.

**Reply by PengWei97** [reply]
But I have a new problem and look forward to your guidance. Before asking my question, I would like to do my best to show my understanding of grain boundary energy or grain boundary mobility modeling based on grain boundary misorientation. My idea of modifying `GBAnisotropy` is to set mob_ij according to the misorientation, and the same for sigma_ij.

When only considering the grain boundary mobility anisotropy, for a grain boundary containing only two order parameters, I use the order parameter mobility L after `GBAnisotropy` as shown in the figure,
![image](https://user-images.githubusercontent.com/56183775/162615423-4b64697e-73f7-41bf-b7c2-6162c03fba0e.png)

Using the same conditions as above, I use the `GBAnisotropyGrainGrowth` written by myself. Its principle is to traverse whether there are two grain IDs that are not equal to `FeatureFloodCount::invalid_id` for each elem. If so, set delta_euler to 3 , otherwise set to 0. The result is shown in the figure,
![image](https://user-images.githubusercontent.com/56183775/162616528-4629c762-a8ce-4d03-af3a-540aaccf218b.png)
And, I plot the two presented results on a graph as follows,
![image](https://user-images.githubusercontent.com/56183775/162627547-29ccb510-673b-42d0-a54e-9d49f120f1ee.png)

I found that L covers much wider than delta_theta  in the grain boundary region (identified in the picture). This is because only one grain ID is valid in the area identified by the red double arrow, so set delta_theta to 0 in `GBAnisotropyGrainGrowth`.

**So the question now is whether it is valid to have two grain IDs at the same time in the area marked by the red snake arrow, instead of only one grain ID being valid at present.**  This makes it possible to make the range of misorientation effect effective in the entire grain boundary region.

The above are the problems I have encountered so far, any suggestions or recommendations to fix these problems would be greatly appreciated.

wei

**Reply by GiudGiud** [reply]
@laagesen

**Reply by laagesen** [reply]
When you say "is it valid to have two grain IDs at the same time..." do you mean you want to change the behavior of how the grain ID's are assigned?

**Reply by PengWei97** [reply]
Hi @laagesen,  this problem was solved according to dschwen's suggestion.

> When you say "is it valid to have two grain IDs at the same time..." do you mean you want to change the behavior of how the grain ID's are assigned?

My real need is to extend the range of effective grain IDs in the grain boundary region. It is obtained by modifying `connecting_threshold`.
